### ISSUE TITLE
Type hint is not respected using nonlocal variable

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#2757]
**Describe the bug**

When assigning a variable that will be used nonlocally pyright disregards the type hint and narrows the type.

**To Reproduce**

```py
from typing import Mapping, Optional, Tuple

def test_custom_http_options() -> None:
    captured: Optional[Tuple[Tuple[object, ...], Mapping[str, object]]] = None
    reveal_type(captured)  # T: None

    def mock___init__(self: object, *args: object, **kwargs: object) -> None:
        nonlocal captured
        captured = (args, kwargs)
        reveal_type(captured)  # T: tuple[tuple[object, ...], dict[str, object]]

    mock___init__('foo')

    # revealed type should be tuple[tuple[object, ...], dict[str, object]] | None
    reveal_type(captured)  # T: None
```

**Expected behavior**

The revealed type should match the type hint.

**VS Code extension or command-line**

Command-line: 1.1.199

**Comment by erictraut** [comment]
The current behavior is "as designed". The type declaration is being honored, but the type of `captured` is narrowed on assignment to `None` within the scope of the `test_custom_http_options` function. Pyright makes the assumption that a symbol that is used within one execution context isn't modified in another context. This is true for globals, nonlocals, instance variables, etc. Mypy makes similar assumptions. The thinking is that the benefit of performing assignment-based type narrowing outweighs the downside of perhaps making a bad assumption about these values being modified "behind the back" by another execution context.

For example:
```python
my_global: int | None

def func1(x: dict[str, str], y: list[str]):
    global my_global
    my_global = 3
    x["a"] = "a"
    y[0] = "a"
    
    # func2 modifies the values of my_global, x['a'] and y[0].
    func2(x, y)

    reveal_type(my_global) # Literal[3]
    reveal_type(x["a"]) # Literal["a"]
    reveal_type(y[0]) # Literal["a"]


def func2(x: dict[str, str], y: list[str]):
    global my_global
    my_global = None
    x["a"] = "b"
    y[0] = "b"

```

I'll think about this a bit more, but we're likely to conclude that this behavior is intended, and you should avoid doing this "behind the back" modification of variables if you want static type checking to work.

**Comment by RobertCraigie** [comment]
Ah I see, that makes sense. I was originally confused to why there was type narrowing applied at declaration time but I can understand the use cases for this.

I wonder if it would be worth special casing nonlocal usage as I think that is a fairly common pattern, as long as there are no performance regressions / increased complexity by doing so as there is a simple workaround:

```py
captured = cast(Optional[Tuple[Tuple[object, ...], Mapping[str, object]]], None)
```

**Comment by erictraut** [comment]
The problem is that within the `test_custom_http_options` method, there is no `nonlocal` access. From the perspective of this function, `captured ` is simply a local variable. The `nonlocal` access is in `mock___init__`, which is "reaching in" to the scope of the other function and modifying its local variables. 

Even if the nonlocal access was within the outer function, excluding it from narrowing on assignment would be problematic and lead to other false positive errors.

I think this is just an "as designed" behavior. Changing it will make the rules inconsistent and will do nothing but move false positives from one place to another, so that's not a good tradeoff. I'm going to therefore mark this as "won't fix".
